### PR TITLE
opal/accelerator: Add IPC APIs

### PR DESCRIPTION
**Pull request by wckzhang**
Signed-off-by: William Zhang <wilzhang@amazon.com>

### CODE REVIEW

**Comment by wckzhang** [comment]
A start for adding generic IPC functionality.

File: opal/mca/accelerator/accelerator.h [hunk 1]
```diff
@@ -309,6 +309,71 @@ typedef int (*opal_accelerator_base_module_mem_release_fn_t)(
...
+ * @param[IN] dev_ptr        Base pointer to previously allocated device
+ *                           memory
+ * @param[OUT] handle        Pointer to user allocated mem handle to return
+ *                           the memory handle in.
+ *
+ * @return                   OPAL_SUCCESS or error status on failure
+ *
+ */
+typedef int (*opal_accelerator_base_module_get_ipc_handle_fn_t)(
+    int dev_id, void *dev_ptr, void **handle);
```
Comment (bwbarrett) [inline opal/mca/accelerator/accelerator.h:349]: Why a void* instead of a class object like the rest of the interface?

**Comment by wenduwan** [comment]
@edgargabriel Reminded me of this PR. I shall pick it up soon.

File: opal/mca/accelerator/accelerator.h (hunk 1 above)
Comment (wenduwan) [reply opal/mca/accelerator/accelerator.h]: New revision encapsulates `handle` and `size`, therefore we don't need the API to get handle size.

However, I think it's necessary to keep device id and base address - didn't find a simplistic way to hide them, maybe in IPC cache if we decide to implement that.

File: opal/mca/accelerator/accelerator.h [hunk 2]
```diff
@@ -111,6 +111,13 @@ struct opal_accelerator_stream_t {
 };
 typedef struct opal_accelerator_stream_t opal_accelerator_stream_t;
 
+#define MAX_IPC_HANDLE_SIZE 64
```
Comment (edgargabriel) [inline opal/mca/accelerator/accelerator.h:114]: I double checked, hipIpcMemHandle_t is 64 bytes, so this size is large enough, but there is also no room for error. I think at the minimum we want to confirm that 64 bytes are sufficient for cuda and ze as well.

Comment (wenduwan) [reply opal/mca/accelerator/accelerator.h:114]: Thanks for checking. I copied this from libfabric https://github.com/ofiwg/libfabric/blob/main/include/ofi.h#L274

It's also used by ze so I guess it's fine. Also this var is internal to opal so we can extend it if need.

**Review by edgargabriel — APPROVED** [review]
LGTM

**Comment by edgargabriel** [comment]
Since @bwbarrett did the initial review of this pr, it might be good to get his review also in

**Comment by bwbarrett** [comment]
> Since @bwbarrett did the initial review of this pr, it might be good to get his review also in

I'm going to be under water with things to do the next couple of weeks, so don't wait for my review.

**Comment by wenduwan** [comment]
@lrbison Since Brian is down, could you give it a review?

**Review by lrbison — APPROVED** [review]
This is a pretty short PR.  I am not tracking the IPC functions well enough to know if this is sufficient for all use cases.  However I don't see anything wrong with it and it shouldn't break existing code so I see no problem in approving it and getting it merged either.

**Comment by edgargabriel** [comment]
@wenduwan as I mentioned in the meeting, I think meanwhile that we are missing some functions. Assuming that the goal is to be able to remove the remaining hard-coded calls to cuda and replace them with accelerator framework API function calls, here are the functions that are still missing. Not all of them are IPC related, so not sure whether they would absolutely have to be part of this PR:

```
IpcOpenEventHandle
IpcGetEventHandle
StreamWaitEvent

EventDestroy
```
In addition, looking over the current Accelerator API, I noticed that we have an interface to create a stream, but not to destroy a stream ( similarly on how we could create an event, but not destroy one, that function is however already in the list above).

**Comment by wenduwan** [comment]
@edgargabriel Thanks for the reminder - I took a second look at [CUDA](https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__MEM.html#group__CUDA__MEM_1gf1d525918b6c643b99ca8c8e42e36c2e) and [ZE](https://spec.oneapi.io/level-zero/latest/core/api.html) API docs and added a couple more that I think are important. Please let me know if that makes sense to you.

I do not have extensive experience with accelerators so I'm open to suggestions to remove less common APIs.

Also I want to mention that I didn't see a `CloseIpcEventHandle` API - should I add one?

**Comment by edgargabriel** [comment]
> Also I want to mention that I didn't see a `CloseIpcEventHandle` API - should I add one?

@wenduwan good point. As far as I can see, neither hip nor cuda have a `closeIpcEventHandle` function, and they don't even mention in the documentation what to do with the handle. Does ze have a function for that purpose? I will also check with our hip team what is the recommended action.

Since I just read through the documentation of the ipc functionality, I am also double checking one more aspect, namely the fact that the events for IPC have to be opened with certain flags set (`[cuda/hip]EventDisableTiming` and `[cuda/hip]EventInterprocess`). I *think* we might need another API function for that (e.g. `create_event_for_ipc()` or similar). I will come back to you on that.

**Comment by edgargabriel** [comment]
@wenduwan I would suggest to go ahead and merge the PR as is, i.e. without the closeIpcEventHandle. There is no hip or cuda function along those lines, and because of that its also not used anywhere in the code. If it turns out later that we still need one for another accelerator, we can still add the API in, we do have a bit of time before the 5.1 release

**Comment by wenduwan** [comment]
@edgargabriel Cool thanks for double checking.

@hppritcha Since you chimed in on this during the Tuesday call I want to see if you have other comments?

**Review by hppritcha — APPROVED** [review]

**Comment by wenduwan** [comment]
@lrbison @hppritcha Can we merge this? I will open backport to 5.0

**Comment by edgargabriel** [comment]
@wenduwan why would you backport this to 5.0? I think this is meant for 5.1, its new functionality?

**Comment by wenduwan** [comment]
> why would you backport this to 5.0? I think this is meant for 5.1, its new functionality?

I see. Cool then we can leave it on main.

**Review by lrbison — APPROVED** [review]

**Pull request merged**